### ISSUE TITLE
Infinity Underflow in Cross-Entropy with Softmax

### ISSUE THREAD

**Issue by dkmisra** [deeplearning4j/deeplearning4j#785]
## Bug

When using weighted-softmax with cross-entropy one gets a scenario as:

` Loss(p,q) = - \sum_i p_i log q_i ` where p is the target and q is the output of softmax

when the target is a one-hot vector with label l we get:

`Loss(p,q) =  - log q_label`

this is same as negative log-likelihood (NLL) loss-function.  Now consider the weight softmax from nin units to nout units (nout being the dimensionality of p,q). We get:

> `q_label = e^{\sum_{i=1}^nin W_{i,label}a_i + b_{label} } / \sum_k e^{\sum_{i=1}^nin W_{i,k} a_i + b_k}`

now even if activations are weights are bounded in -1 to 1. One can still get a very low exponent if `nin` is very large. Consider `nin` to be 100 and all weights to be -1 and activations as 1. We get the numerator in the softmax as 

`e^{-1 -1 -1 .... -1 (101 times)} = e^-101`

this can start underflowing thus giving us log(0) term in the loss function and killing the learning.

## Suggestion:

Use log in combination with softmax instead of first computing the softmax term and then applying the log. 

## Update:

I know this has been resolved in the next release. But I feel that I should still file this issue.

**Comment by AlexDBlack** [comment]
Should have been solved as of https://github.com/deeplearning4j/nd4j/pull/458

**Comment by hrxx** [comment]
I met this problem, thank you!
